### PR TITLE
[Intel MKL] Add weight cache for FP32 MatMul.

### PR DESCRIPTION
**Pull request by Zantares**
This PR will cache weight for FP32 MatMul if it's const. Also move some common code related to cache weight from `mkl_qmatmul_op.cc` to `mkl_matmul_ops_common.h`.

Modified:
- tensorflow/core/graph/mkl_layout_pass.cc
- tensorflow/core/kernels/mkl_fused_ops_test.cc
- tensorflow/core/kernels/mkl_matmul_op_fused.cc
- tensorflow/core/kernels/mkl_matmul_ops_common.h
- tensorflow/core/kernels/mkl_qmatmul_op.cc
- tensorflow/core/ops/mkl_nn_ops.cc

Signed-off-by: Lu Teng teng.lu@intel.com

### CODE REVIEW

File: tensorflow/core/kernels/mkl_matmul_ops_common.h [hunk 1]
```diff
@@ -374,9 +374,90 @@ class MklDnnMatMulOpBase : public OpKernel {
...
+      OpKernelContext* context,
+      const std::shared_ptr<mkldnn::inner_product_forward::primitive_desc>&
+          matmul_fwd_pd,
+      Tweight* weight_data, const Tensor& weight_tensor,
+      MklDnnData<Tweight>& weight, const memory::desc& weight_md)
+      LOCKS_EXCLUDED(mu_) {
+    mutex_lock lock(mu_);
+    const Tensor& weight_t = *weight_oi.AccessTensor(context);
+
+    // if the weights are already cahced, there's nothing to do
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_ops_common.h]: Nit:
```suggestion
    // If the weights are already cached, there's nothing to do
```

File: tensorflow/core/kernels/mkl_matmul_ops_common.h [hunk 2]
```diff
@@ -374,9 +374,90 @@ class MklDnnMatMulOpBase : public OpKernel {
...
+    }
+    return nullptr;
+  }
+
   engine cpu_engine_ = engine(engine::cpu, 0);
 
  protected:
+  // Tensor to save reordered weight
+  mutex mu_;
+  PersistentTensor weight_oi GUARDED_BY(mu_);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_ops_common.h]: Nit: class member variable names should end with underscore.
```suggestion
  PersistentTensor weight_oi_ GUARDED_BY(mu_);
```

File: tensorflow/core/kernels/mkl_matmul_ops_common.h [hunk 3]
```diff
@@ -374,9 +374,90 @@ class MklDnnMatMulOpBase : public OpKernel {
...
+  PersistentTensor weight_oi_md GUARDED_BY(mu_);
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_matmul_ops_common.h]: Nit: Same here.
```suggestion
  PersistentTensor weight_oi_md_ GUARDED_BY(mu_);
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for my delay! I have minor comments.

**Comment by gbaned** [comment]
@Zantares Could you please check reviewer comments and keep us posted. Thanks!

**Comment by Zantares** [comment]
Hi @gbaned , thanks for your reminder. Because of the well-known virus in China, my work will be delayed in 1~2 weeks. I'll refine this PR as soon as I get back to work, thanks.

**Comment by penpornk** [comment]
@Zantares Please take your time and I wish you all the best!

File: tensorflow/core/kernels/mkl_matmul_ops_common.h (hunk 3 above)
Comment (Zantares) [reply tensorflow/core/kernels/mkl_matmul_ops_common.h]: Done.

File: tensorflow/core/kernels/mkl_matmul_ops_common.h (hunk 2 above)
Comment (Zantares) [reply tensorflow/core/kernels/mkl_matmul_ops_common.h]: Done.

**Comment by Zantares** [comment]
Hi @penpornk , please have a look at the modified code, thanks!

**Review by penpornk — APPROVED** [review]
Thank you again for the PR!

**Pull request merged**